2_Code/src/uart.c

```diff
@@ -1,20 +1,95 @@
+/* 
+ * Date: 03/30/2024
+ * Course number: EE 2361
+ * Term: Spring
+ * Lab/assignment number: Final Project
+ * Short Program Description: UART Library to write and read to Grove - GPS Air530.
+ */
+
 #include "xc.h"
-#include "uart.h"
+#include "UART.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+volatile unsigned char buffer[64]; // Circular buffer to store received data
+volatile unsigned char front = 0; // Index for adding data to front of the buffer
+volatile unsigned char back = 0; // Index for reading data from the back of the buffer
+
+void __attribute__((__interrupt__,__auto_psv__)) _U1RXInterrupt(void)
+{
+    // Updates the circular buffer with received UART data
+    // Clears the UART receive interrupt flag
+    // front: index for storing data in the buffer
+    // U1RXREG: received data from UART
+    
+    IFS0bits.U1RXIF = 0;
+    buffer[front++] = U1RXREG;
+    front &= 63;
+}
+
+void send_GPS_Str_command(unsigned char* s)
+{
+    // Sends a String to the GPS module using send_GPS_Char_command function
+    // s: the string to be transmitted
+    
+    int i = 0;
+    while (s[i] != '\0') 
+    {
+        send_GPS_Char_command(s[i]);
+        i++;
+    }
+    
+}
 
-void initUART(int baudRate)
+void send_GPS_Char_command(unsigned char command)
 {
-    // Set pins to digital in/out
+    // Sends a character to the GPS module via UART
+    // command: the character to be transmitted
+
+    while (!IFS0bits.U1TXIF);
+    IFS0bits.U1TXIF = 0;
+    U1TXREG = command;
+}
 
-    // Set up UART (not complete)
-    U1MODE = 0x0;        // 8-bit, no parity, 1 stop bit
-    U1MODEbits.BRGH = 0; // Not high speed
-    U1BRG = (unsigned int)FCY / (16 * baudRate) - 1; // Calculate baud (should be 103)
-    U1MODEbits.ON = 1;   // Enable UART
+unsigned char get_GPS_data(void)
+{
+    // Retrieves a single NMEA character from the circular buffer
+    // Waits until data is available (front is different from back)
+    // NMEA: the retrieved character
+    
+    unsigned char NMEA;
 
-    // Peripheral Pin Select
+    while (front == back) ;
+    NMEA = buffer[back++];
+    back &= 63;
```

Comment:
Cool algorithm, but for our understanding, could you clarify why we use `back &= 63;`?
Thanks!